### ISSUE TITLE
Question - Pattern for receiving messages from more than one queue

### ISSUE THREAD

**Issue by DaveSenn** [Azure/amqpnetlite#448]
I've a requirement for an application to consume messages from multiple (~10) queues. When searching for this I was not able to find any examples showing how to consume messages from multiple queues.
I've to work with ArtemisMQ. It is my understanding that it is recommended to use a single connection for all interactions with the broker. This would mean that i've to create multiple receiverlinks on the same connection and itterate over them to try to receive a message. But somehow this feals like there should be a better solution to this problem.

I've created a small demo application to show my implementation (you can find the whole code below).
The "interesting" part looks like this:

```
while ( !_cancellationToken.IsCancellationRequested )
	foreach ( var receiver in receivers )
	{
		// ReceiveAsync( TimeSpan.Zero ); blocks forever and no messages will be received 
		var message = await receiver.ReceiveAsync( TimeSpan.FromMilliseconds( 1 ) );
		if ( message == null )
			continue;

		receiver.Accept( message );
	}
```

Is there a better way to achieve this than looping over multiple receivers and call `ReceiveAsync` with a very low timeout?

```
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Amqp;
using Amqp.Framing;
using Amqp.Types;

namespace MultipleReceiverDemo
{
    public sealed class Program
    {
        private const String Address = "amqp://admin:password@localhost:5672";

        public static async Task Main()
        {
            Console.WriteLine( "Starting producer/consumer demo... press <enter> to stop." );
            var cts = new CancellationTokenSource();

            var producerA = new MessageProducer( Address, "producerA", "queue-a", cts.Token );
            var producerB = new MessageProducer( Address, "producerB", "queue-b", cts.Token );
            var producerATask = producerA.StartSendingMessages();
            var producerBTask = producerB.StartSendingMessages();

            var messageConsumer = new MessageConsumer( Address, "myConsumer", new[] { "queue-a", "queue-b" }, cts.Token );
            var receivingTask = messageConsumer.StartReceivingMessages();

            Console.ReadLine();
            cts.Cancel();
            await Task.WhenAll( producerATask, producerBTask, receivingTask );
        }
    }

    public sealed class MessageConsumer
    {
        private readonly String _address;
        private readonly CancellationToken _cancellationToken;
        private readonly String _consumerName;
        private readonly String[] _destinations;

        public MessageConsumer( String address, String consumerName, String[] destinations, CancellationToken cancellationToken )
        {
            _address = address;
            _consumerName = consumerName;
            _destinations = destinations;
            _cancellationToken = cancellationToken;
        }

        public async Task StartReceivingMessages()
        {
            await Task.Yield();
            while ( !_cancellationToken.IsCancellationRequested )
            {
                var connectionFactory = new ConnectionFactory();
                var address = new Address( _address );
                try
                {
                    var connection = await connectionFactory.CreateAsync( address );
                    var session = ( (IConnection) connection ).CreateSession();
                    var receivers = new List<IReceiverLink>();
                    foreach ( var destination in _destinations )
                    {
                        var receiver = session.CreateReceiver( $"{_consumerName}_{destination}",
                                                               new Source
                                                               {
                                                                   Address = destination,
                                                                   Capabilities = new[] { new Symbol( "queue" ) }
                                                               } );
                        receivers.Add( receiver );
                    }

                    while ( !_cancellationToken.IsCancellationRequested )
                        foreach ( var receiver in receivers )
                        {
                            // ReceiveAsync( TimeSpan.Zero ); blocks forever and no messages will be received 
                            var message = await receiver.ReceiveAsync( TimeSpan.FromMilliseconds( 1 ) );
                            if ( message == null )
                                continue;

                            receiver.Accept( message );
                            Console.WriteLine( $"{_consumerName} - Received message with id: '{message.Properties.MessageId}'" );
                        }
                }
                catch ( Exception ex )
                {
                    Console.WriteLine( $"{_consumerName} - Connection error in producer '{_consumerName}' {ex.Message} => create new connection." );
                    await Task.Delay( 1000, CancellationToken.None );
                }
            }
        }
    }

    public sealed class MessageProducer
    {
        private readonly String _address;
        private readonly CancellationToken _cancellationToken;
        private readonly String _destination;
        private readonly String _producerName;

        public MessageProducer( String address, String producerName, String destination, CancellationToken cancellationToken )
        {
            _address = address;
            _producerName = producerName;
            _destination = destination;
            _cancellationToken = cancellationToken;
        }

        public async Task StartSendingMessages()
        {
            while ( !_cancellationToken.IsCancellationRequested )
            {
                var connectionFactory = new ConnectionFactory();
                var address = new Address( _address );
                var messageCounter = 0;
                try
                {
                    var connection = await connectionFactory.CreateAsync( address );
                    var session = ( (IConnection) connection ).CreateSession();
                    var sender = session.CreateSender( _destination,
                                                       new Target
                                                       {
                                                           Address = _destination,
                                                           Capabilities = new[] { new Symbol( "queue" ) }
                                                       } );

                    while ( !_cancellationToken.IsCancellationRequested )
                    {
                        var message = new Message( $"Message #{messageCounter} from {_producerName}." )
                        {
                            Properties = new()
                            {
                                MessageId = $"{_producerName} #{messageCounter}"
                            }
                        };

                        await sender.SendAsync( message );
                        Console.WriteLine( $"\t{_producerName} - Sent message with id: '{message.Properties.MessageId}'" );

                        await Task.Delay( 1000, CancellationToken.None );
                        messageCounter++;
                    }
                }
                catch ( Exception ex )
                {
                    Console.WriteLine( $"{_producerName} - Connection error in producer '{_producerName}' {ex.Message} => create new connection." );
                    await Task.Delay( 1000, CancellationToken.None );
                }
            }
        }
    }
}
```

**Comment by Havret** [comment]
You can take a look at this repo --> https://github.com/Havret/dotnet-activemq-artemis-client
and this class in particular https://github.com/Havret/dotnet-activemq-artemis-client/blob/master/src/ActiveMQ.Artemis.Client/Consumer.cs

Here is a short article that describes how to use the lib in ASP.NET Core environment. --> https://havret.io/activemq-artemis-net-core

**Comment by DaveSenn** [comment]
Hi @Havret thank you very much for your answer.
I did read your blog post and take a look at your code. It was very interesting for me.

Do I understand this correctly, your code creates a `Consumer`/`ReceiverLink` for each call to `AddConsumer` .
For each created consumer your code asynchronously waits for incoming messages:

```
 _task = Task.Run(async () =>
            {
                while (!token.IsCancellationRequested)
                {
                        var msg = await consumer.ReceiveAsync(token).ConfigureAwait(false);
                }
            }, token);
```

But `consumer.ReceiveAsync` does not actually await a method on the underlaying `ReceiverLink` but waits for new messages on an internal `Channel`.
Your code subscribes to an event/callback `Start` defined on the `ReceiverLink` to receive messages and add them to the internal `Channel`.

So you would recommend me to use the callback passed to `Start` to receive messages from multiple `ReceiverLink`s?

**Comment by Havret** [comment]
Exactly. Just be mindful not to do processing inside of `Start`s callback. It will block the connection pump if you do there any blocking / time consuming processing. That's why I use Channels to offload this processing somewhere else.

**Comment by pm9999** [comment]
Hi @Havret, I did read your post and tried your AMQ solution. It is well detailed and very interesting.

https://havret.io/activemq-artemis-net-core
But it does not create durable consumers, right? I tried the above approach but it creates queues only when the Consumer Application is running. Do you know how to create durable consumers while using ArtemisNetClient library? or amqnetlite. I could not find much help on the other blogs. Thank you!

**Comment by Havret** [comment]
Hi @pm9999, thank you for your kind words. At the moment you need to use the management API via `ITopologyManager` interface to pre-create (or pre-declare) queues, and addresses to have the full-blown shared durable consumer. Here you can find a simple code snippet that demonstrates how it can be done:
- https://github.com/Havret/dotnet-activemq-artemis-client/blob/0292ce05da70f9bd7e647630f81d54b9ef6cee53/test/ArtemisNetClient.IntegrationTests/AddressModelSpec.cs#L89-L142

In general, this approach is very similar to what you can find in RabbitMQ client. However, for the vast majority of the use cases, I came across so far, there was no need to use this API. Most organizations treat their queue topology with great care, especially in terms of durable queues. As a result, the broker ships with pre-created configuration, and the clients simply attach to the queues that already exist. 

The other option is to create a shared durable consumer as so-called shared durable subscription (8.7 https://www.oasis-open.org/committees/download.php/67638/amqp-bindmap-jms-v1.0-wd10.pdf). This is something I'm considering adding to the ArtemisNetClient (https://github.com/Havret/dotnet-activemq-artemis-client/issues/312) alongside with shared non-durable consumers (https://github.com/Havret/dotnet-activemq-artemis-client/issues/295). Having these features you could achieve similar behavior as with `ITopologyManager`, but without the need to expose management API to the clients. If you think that this might be something useful to you, feel free to reach me out on ArtemisNetClient repo.

**Comment by pm9999** [comment]
Thank you for the response @Havret. I will look into the ITopologyManager interface.
Yes, it would be great if creation of shared durable subscriptions is added to the ArtemisNetClient!

**Comment by Havret** [comment]
@pm9999 The features are released with version 2.5.0. You can give them a try. https://github.com/Havret/dotnet-activemq-artemis-client/releases/tag/v2.5.0

**Comment by pm9999** [comment]
> @pm9999 The features are released with version 2.5.0. You can give them a try. https://github.com/Havret/dotnet-activemq-artemis-client/releases/tag/v2.5.0

Thank you so much @Havret! I will definitely give this a try!
In the meanwhile we used amqpnetlite package and developed the solution.
